lib/app.rb

```diff
@@ -0,0 +1,99 @@
+# rubocop:disable Style/CaseEquality
+module Enumerable
+  def my_each
+    return enum_for unless block_given?
+
+    list = is_a?(Range) ? to_a : self
+    list.length.times do |i|
+      yield(list[i])
+    end
+    self
+  end
+
+  def my_each_with_index
+    return enum_for unless block_given?
+
+    list = is_a?(Range) ? to_a : self
+    list.length.times do |i|
+      yield(list[i], i)
+    end
+    self
+  end
+
+  def my_select
+    return enum_for unless block_given?
+
+    selected = []
+    my_each do |el|
+      selected << el if yield(el)
+    end
+    selected
+  end
+
+  def my_all?(arg = nil)
+    if arg
+      my_each { |ele| return false unless arg === ele }
+    elsif block_given?
+      my_each { |ele| return false unless yield(ele) }
+    else
+      my_each { |ele| return false unless ele }
+    end
+    true
+  end
+
+  def my_any?(arg = nil)
+    if arg
+      my_each { |ele| return true if arg === ele } # rubocop:disable Style/CaseEquality
+    elsif block_given?
+      my_each { |ele| return true if yield(ele) }
+    else
+      my_each { |ele| return true if ele }
+    end
+    false
+  end
+
+  def my_none?(arg = nil, &block)
+    !my_any?(arg, &block)
+  end
+
+  def my_count(arg = nil, &prc)
+    counted = 0
+    if block_given?
+      my_each { |ele| counted += 1 if prc.call(ele) }
+    elsif !block_given? && arg.nil?
+      counted = to_a.length
+    else
+      counted = to_a.my_select { |ele| ele == arg }.length
+    end
+    counted
+  end
+
+  def my_map(arg = nil, &prc)
+    return enum_for unless block_given?
+
+    list = is_a?(Range) ? to_a : self
+    mapped = []
+    if arg.nil?
+      list.my_each { |ele| mapped << yield(ele) }
+    else
+      list.my_each { |ele| mapped << prc.call(ele) }
+    end
+    mapped
+  end
```

Comment:
- [ ]  Executes only the proc when both a proc and a block is given and returns a **new array** with the results of the proc